pkg/summarizer/enhancedanalyzer/analyzer.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2020 The TestGrid Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// An extension of naiveanalyzer with an improved algorithm
+package enhancedanalyzer
+
+import (
+	statepb "github.com/GoogleCloudPlatform/testgrid/pb/state"
+	summarypb "github.com/GoogleCloudPlatform/testgrid/pb/summary"
+	"github.com/GoogleCloudPlatform/testgrid/pkg/summarizer/baseanalyzer"
+	"github.com/GoogleCloudPlatform/testgrid/pkg/summarizer/common"
+)
+
+const analyzerName = "enhancedanalyzer"
+
+// IntString is for sorting, primarily intended for map[string]int as implemented below
+type IntString struct {
+	s string
+	i int
+}
+
+// EnhancedAnalyzer implements functions that calculate flakiness as a ratio of failed tests to total tests
+type EnhancedAnalyzer struct {
+}
+
+// GetFlakiness returns a HealthinessInfo message
+func (ea *EnhancedAnalyzer) GetFlakiness(gridMetrics []*common.GridMetrics, relevantFilteredStatus map[string][]statepb.Row_Result, minRuns int, startDate int, endDate int, tab string) *summarypb.HealthinessInfo {
+	// Delegate to a BaseAnalyzer and change the flakiness scores.
+	// (And average flakiness)
+	na := baseanalyzer.BaseAnalyzer{}
+	hi := na.GetFlakiness(gridMetrics, relevantFilteredStatus, minRuns, startDate, endDate, tab)
+	averageFlakiness := float32(0)
+	for _, test := range hi.Tests {
+		test.Flakiness = calculateEnhancedFlakiness(relevantFilteredStatus[test.DisplayName])
+		averageFlakiness += test.Flakiness
+	}
+	hi.AverageFlakiness = averageFlakiness / float32(len(hi.Tests))
+	return hi
+}
+
+const IGNORE_FAILURES_IN_A_ROW = 3
+
+func consecutiveFailures(statuses []statepb.Row_Result, i int) int {
+	result := 0
+	for i < len(statuses) && statuses[i] == statepb.Row_FAIL {
+		result++
+		i++
+	}
+
+	return result
+}
+
+// calculateEnhancedFlakiness gets a calculation of flakiness based on number of flips to failing rather than number of failures
+// statuses should have already filtered to the correct time horizon and removed infra failures
+// Returns a percentage between 0 and 100
+func calculateEnhancedFlakiness(statuses []statepb.Row_Result) float32 {
+	flips := 0
+	considered := 0
+	lastPassing := true // No flakes if we pass 100%
+	i := 0
+	for i < len(statuses) {
+		cf := consecutiveFailures(statuses, i)
+		if cf >= IGNORE_FAILURES_IN_A_ROW {
+			// Ignore the run of failures
+			i += cf
+			if i >= len(statuses) {
+				break
+			}
+		}
+		s := statuses[i]
+		considered++
+		if s == statepb.Row_PASS {
```

Comment:
This set of statuses (PASS, FLAKY, other) won't be sufficient since multiple statuses can be 'passing' and multiple statuses can be 'failing', though it might be okay for particular dashboards. I'd either fix that, or note that in the comments here and add a TODO linked to https://github.com/GoogleCloudPlatform/testgrid/issues/186 to properly check what's passing vs. failing.

Reply:
I changed this to be a specific set of values that exists only here.

A bit awkward, but at least it's not misleading.

Reply:
Yeah, the statuses here come from results.Coalesce and other filtering in the new code. Whether the values from Coalesce are right is another story, but I think they should be pretty similar and if need be, Coalesce can be changed or we can use a different function that works similarly and covers everything correctly. I think for now something helpful for clarity would be to add a comment mentioning that these values have been filtered previously. Also this should be a switch case statement for consistency with other parts of the code.